WordPress/Classes/ViewRelated/Tools/SettingsTextViewController.m

```diff
@@ -126,16 +226,59 @@ - (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger
 }
 
 
-#pragma mark - UITextFieldDelegate Methods
+#pragma mark - Helpers
 
-- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
+- (void)dismissViewController
 {
-    NSRange newLineRange = [string rangeOfCharacterFromSet:[NSCharacterSet newlineCharacterSet]];    
-    if (newLineRange.location != NSNotFound) {
+    if (self.isModal) {
+        [self dismissViewControllerAnimated:YES completion:nil];
+    } else {
         [self.navigationController popViewControllerAnimated:YES];
     }
+}
+
+- (void)notifyValueDidChangeIfNeeded
+{
+    if (self.onValueChanged == nil || [self.textField.text isEqualToString:self.text]) {
+        return;
+    }
+    
+    self.onValueChanged(self.textField.text);
+    self.onValueChanged = nil;
```

Comment:
> onValueChanged should only be called once

I see no documentation that says that :wink: :wink: 

> This is meant to prevent double call, in the scenario in which the VC is in a NavigationController stack, and gets dismissed thru the keyboard

I think the problem here is that you don't need to call this method from `doneButtonWasPressed:` since it's already handled when the VC disappears. I've never really liked that pattern of calling it on disappear, and I feel it's only getting worse. I'm open to ideas on how to design this better.

I've also just noticed that you're setting this to `nil` on `cancelButtonWasPressed:` and it doesn't feel right either.

While the common use case is that you'll present this and get a callback once when it's dismissed (either via Done or popping back), the current implementation doesn't really guarantee that. For instance, something else in the app could present a modal view controller on top of this (let's say, for re-authentication), `viewWillDisappear:` gets called and it sets `onValueChanged = nil`. When the user is done with the modal and comes back to keep editing, changes are ignored because the callback is set to `nil`.

For the cancel button case, I have trouble imagining a specific scenario where this breaks, but nothing in the documentation implies that a `SettingsTextViewController` can't be reused and presented twice.